File: 2025.07/2025.07.04/3307-find-the-k-th-character-in-string-game-ii.js
/**
 * @param {number} k
 * @param {number[]} operations
 * @return {character}
 */
var kthCharacter = function (k, operations) {
  // Solution: Binary Search + Count
  // Time: O(logK)
  // Space: O(1)

  let n = Math.pow(2, operations.length);

  if (k > n) {
    return "?";
  }

  n = n / 2;
  let count = 0;

  for (let i = operations.length - 1; i >= 0; i--) {
    if (k === 1) {
      // ready to stop since operations won't be used anymore

      break;
    }

    if (n < k) {
      k -= n;
      count += operations[i];
    }

    n = n / 2;
  }

  return String.fromCharCode(97 + (count % 26));
};

console.log(kthCharacter(3, [0, 0, 0])); // 'a'
console.log(kthCharacter(10, [0, 1, 0, 1])); // 'b'

// aa
// aabb
// aabbaabb
// aabbaabbbbccbbcc
//  1       0

// index 10 = index 2 + 1 => index 1 + 1 => 1 => 'b'
// 8 => 4 => 2 => 1

console.log(kthCharacter(2, [1])); // 'b'

// a
// ab
//  0

// index 2 = index 1 + 1 = 1 => 'b'
// 1

console.log(kthCharacter(3, [1])); // '?'

File: 2025.07/2025.07.04/3307-find-the-k-th-character-in-string-game-ii.py
from typing import List

class Solution:
  def kthCharacter(self, k: int, operations: List[int]) -> str:
    # Solution: Binary Search + Count
    # Time: O(logK)
    # Space: O(1)
    
    n = 2 ** len(operations)
    if k > n:
      return '?'
    n = n / 2
    count = 0
    for i in range(len(operations) - 1, -1, -1):
      if k == 1:
        break
      if n < k:
        k -= n
        count += operations[i]
      n /= 2

    return chr(ord('a') + count % 26)

print(Solution().kthCharacter(3, [0, 0, 0]))  # 'a'
print(Solution().kthCharacter(10, [0, 1, 0, 1])) # 'b'
print(Solution().kthCharacter(2, [1])) # 'b'
print(Solution().kthCharacter(3, [1])) # '?'
